lib/tbot/config/output_spiffe_svid.go

```diff
@@ -0,0 +1,276 @@
+	// SVID to help them identify it.
+	Hint string `yaml:"hint,omitempty"`
+	// SANS is the Subject Alternative Names that are requested to be included
+	// in the SVID.
+	SANS SVIDRequestSANs `yaml:"sans,omitempty"`
+}
+
+func (o *SVIDRequest) CheckAndSetDefaults() error {
+	switch {
+	case o.Path == "":
+		return trace.BadParameter("svid.path: should not be empty")
+	case !strings.HasPrefix(o.Path, "/"):
+		return trace.BadParameter("svid.path: should be prefixed with /")
+	}
+	for i, stringIP := range o.SANS.IP {
+		ip := net.ParseIP(stringIP)
+		if ip == nil {
+			return trace.BadParameter(
+				"ip_sans[%d]: invalid IP address %q", i, stringIP,
+			)
+		}
+	}
+	return nil
+}
+
+// SPIFFESVIDOutput is the configuration for the SPIFFE SVID output.
+// Emulates the output of https://github.com/spiffe/spiffe-helper
+type SPIFFESVIDOutput struct {
+	// Destination is where the credentials should be written to.
+	Destination bot.Destination `yaml:"destination"`
+	SVID        SVIDRequest     `yaml:"svid"`
+}
+
+type spiffeSVIDSigner interface {
+	SignX509SVIDs(ctx context.Context, req *machineidv1pb.SignX509SVIDsRequest, opts ...grpc.CallOption) (*machineidv1pb.SignX509SVIDsResponse, error)
+}
+
+func GenerateSVID(
+	ctx context.Context,
+	signer spiffeSVIDSigner,
+	reqs []SVIDRequest,
+	ttl time.Duration,
+) (*machineidv1pb.SignX509SVIDsResponse, *rsa.PrivateKey, error) {
+	ctx, span := tracer.Start(
+		ctx,
+		"GenerateSVID",
+	)
+	defer span.End()
+	privateKey, err := native.GenerateRSAPrivateKey()
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	pubBytes, err := x509.MarshalPKIXPublicKey(privateKey.Public())
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	svids := make([]*machineidv1pb.SVIDRequest, 0, len(reqs))
+	for _, req := range reqs {
+		svids = append(svids, &machineidv1pb.SVIDRequest{
+			PublicKey:    pubBytes,
+			SpiffeIdPath: req.Path,
+			DnsSans:      req.SANS.DNS,
+			IpSans:       req.SANS.IP,
+			Hint:         req.Hint,
+			Ttl:          durationpb.New(ttl),
+		})
+	}
+
+	res, err := signer.SignX509SVIDs(ctx, &machineidv1pb.SignX509SVIDsRequest{
+		Svids: svids,
+	})
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	return res, privateKey, nil
+}
+
+func (o *SPIFFESVIDOutput) Render(
+	ctx context.Context, p provider, _ *identity.Identity,
+) error {
+	ctx, span := tracer.Start(
+		ctx,
+		"SPIFFESVIDOutput/Render",
+	)
+	defer span.End()
+
+	spiffeCAs, err := p.GetCertAuthorities(ctx, types.SPIFFECA)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	res, privateKey, err := GenerateSVID(
+		ctx,
+		p,
+		[]SVIDRequest{o.SVID},
+		// For TTL, we use the one globally configured.
+		p.Config().CertificateTTL,
+	)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	privBytes, err := x509.MarshalPKCS8PrivateKey(privateKey)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	privPEM := pem.EncodeToMemory(&pem.Block{
+		Type:  "PRIVATE KEY",
```

Comment:
Consider using an existing const (or creating some, if we're trying to keep the tbot binary smaller):

https://github.com/gravitational/teleport/blob/98c455b95ac02dc7ea68c6c9481069f85d12b627/api/utils/keys/privatekey.go#L39

Reply:
a1b68bb5bb